synthetic_data/dataset_generator.py

```diff
@@ -0,0 +1,129 @@
+
+
+def get_ordered_column(
+    data: np.array,
+    type: str,
+    original_format: str = "%B %d %Y %H:%M:%S",
+    order: str = "ascending",
+) -> np.array:
+    """Sort a numpy array based on data type.
+
+    :param data: numpy array to be sorted
+    :type data: np.array
+
+    :return: sorted numpy array
+    """
+    if type == "datetime":
+        date_object = np.array([datetime.strptime(dt, original_format) for dt in data])
+        sorted_datetime = np.sort(date_object)
+        sorted_data = np.array([dt.strftime(original_format) for dt in sorted_datetime])
+    else:
+        sorted_data = np.sort(data)
+    if order == "descending":
+        return sorted_data[::-1]
+    return sorted_data
+
+
+def generate_dataset_by_class(
+    rng: Generator,
+    columns_to_generate: Optional[List[dict]] = None,
+    dataset_length: int = 100000,
+    path: Optional[str] = None,
+) -> pd.DataFrame:
+    """Randomly generate a dataset with a mixture of different data classes.
+
+    :param rng: the np rng object used to generate random values
+    :type rng: numpy Generator
+    :param columns_to_generate: Classes of data to be included in the dataset
+    :type columns_to_generate: List[dict], None, optional
+    :param dataset_length: length of the dataset generated
+    :type dataset_length: int, optional
+    :param path: path to output a csv of the dataframe generated
+    :type path: str, None, optional
+    :param ordered: whether to generate ordered data
+    :type ordered: bool, optional
+
+    :return: pandas DataFrame
+    """
+    gen_funcs = {
+        "integer": random_integers,
+        "float": random_floats,
+        "categorical": random_categorical,
+        "text": random_text,
+        "datetime": random_datetimes,
+        "string": random_string,
+    }
+
+    dataset = []
+    for col in columns_to_generate:
+        data_type_var = col.get("data_type", None)
+        if data_type_var not in gen_funcs:
+            raise ValueError(f"generator: {data_type_var} is not a valid generator.")
+        col_generator_function = gen_funcs.get(data_type_var)
+
+        sort = col.get("ordered", None)
+        if sort in ["ascending", "descending"]:
+            if data_type_var == "datetime" and "date_format_list" in col:
+                dataset.append(
+                    get_ordered_column(
+                        col_generator_function(**col, num_rows=dataset_length, rng=rng),
+                        data_type_var,
+                        col["date_format_list"][0],
+                    )
+                )
+            else:
+                dataset.append(
+                    get_ordered_column(
+                        col_generator_function(**col, num_rows=dataset_length, rng=rng),
+                        data_type_var,
+                    )
+                )
+        else:
+            dataset.append(
+                col_generator_function(**col, num_rows=dataset_length, rng=rng)
+            )
```

Comment:
I think we can clean this up to be more readable.

Think about it more like:
1. Generate code
2. if Ordered, sort the data.

I see the caveat we have for datetime, but maybe a better approach is to alter how we handle datetime generators. If we want a quick solution, we could alter datetime to accept an order parameter and sort within the generator itself. 

That would clean this code to be a single `if `. I don't like that it is doing things differently, but we already are above in the sort by not using the generator for datetime which IMO I don't think is appropriate.

Reply:
yep yep -- good call! @drahc1R definitely glad to talk more about improving this -- making the code DRY-er